src/libcore/fmt/mod.rs

```diff
@@ -259,16 +251,23 @@ struct Void {
 #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
 #[doc(hidden)]
 pub struct ArgumentV1<'a> {
-    value: &'a Void,
-    formatter: fn(&Void, &mut Formatter<'_>) -> Result,
+    value: &'a Opaque,
+    formatter: fn(&Opaque, &mut Formatter<'_>) -> Result,
 }
 
-impl<'a> ArgumentV1<'a> {
-    #[inline(never)]
-    fn show_usize(x: &usize, f: &mut Formatter<'_>) -> Result {
-        Display::fmt(x, f)
-    }
+// This gurantees a single stable value for the function pointer associated with
+// indices/counts in the formatting infrastructure.
+//
+// Note that a function defined as such would not be correct as functions are
+// always tagged unnamed_addr with the current lowering to LLVM IR, so their
+// address is not considered important to LLVM and as such the as_usize cast
+// could have been miscompiled. In practice, we never call as_usize on non-usize
+// containing data (as a matter of static generation of the formatting
+// arguments), so this is merely an additional check.
+#[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
+static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |_, _| loop {};
```

Comment:
cc @eddyb -- am I correct that this reasoning is correct / this should work?

Reply:
I don't see how it's much more different from the previous approach.
LLVM will replace loads from `USIZE_MARKER` with the `fn` pointer, which is still `unnamed_addr`.

Reply:
In my reading of [your comment](https://github.com/rust-lang/rust/issues/54685#issuecomment-425864863) in a different issue, my understanding is that the non-determinism is strictly in the casting from a function (or closure) to the function pointer.

But even if the closure here is unnamed_addr, the function pointer produced is guaranteed to be stable. I guess what we don't guarantee is that this closure did not get merged with a different function. But that's fine, we don't technically need to -- anywhere that this could have gotten merged, would have to be associated with a usize in the "value" part of Argument, right? So the safety property would still hold.

Reply:
Ah, I see what the argument is. Indeed, we shouldn't introduce non-determinism and `USIZE_MARKER == USIZE_MARKER` should always hold.

The funny thing though is that because you used `|_, _| loop {}`, anyone doing the same thing (`loop {}` in the body) could end up being collapsed to this (I don't think the signature matters, because `usize` is not passed by value).

Reply:
The signature not mattering is an interesting point. There's probably no way to prevent that collapsing, though, right? We could plausibly do something weird (e.g., print the address of the passed references shifted to the right or something) and just say "no one else will do this", I guess.

Reply:
I think you should at least load an `usize` and blackbox it.
Otherwise there is no guarantee there *is* an `usize` (or a same-size integer) behind the opaque pointer.

Reply:
I will try to do this in a follow-up PR to avoid getting this bogged down (and r? you on that one).